File: src/test/java/com/checkers/room/RoomServiceTest.java
package com.checkers.room;

import com.checkers.room.impl.RoomServiceImpl;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.InjectMocks;
import org.mockito.junit.jupiter.MockitoExtension;
import org.springframework.test.util.ReflectionTestUtils;

import java.util.Map;

import static org.assertj.core.api.Assertions.as;
import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.Mockito.mock;

@ExtendWith(MockitoExtension.class)
class RoomServiceTest {

    private final RoomService underTest = new RoomServiceImpl();

    private Map<Integer, Room> rooms;

    @SuppressWarnings("unchecked")
    @BeforeEach
    void setUp() {
        rooms = (Map<Integer, Room>) ReflectionTestUtils.getField(underTest, "rooms");
    }

    @Test
    void testFindOrCreateRoomWhileRoomDoesNotExist() {
        // when
        var actual = underTest.findOrCreateRoomById(1);

        // then
        assertThat(actual).isNotNull();
        assertThat(actual.getId()).isEqualTo(1);
        assertThat(rooms.containsKey(1)).isTrue();
    }

    @Test
    void testFindOrCreateRoomWhileRoomExists() {
        // given
        var room = mock(Room.class);
        rooms.put(1, room);

        // when
        var actual = underTest.findOrCreateRoomById(1);

        // then
        assertThat(actual).isSameAs(room);
    }
}

File: app/src/test/kotlin/ru/kiseru/checkers/room/RoomServiceTest.kt
package ru.kiseru.checkers.room

import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import org.mockito.InjectMocks
import org.mockito.Mock
import org.mockito.junit.jupiter.MockitoExtension
import org.mockito.kotlin.eq
import org.mockito.kotlin.given
import org.mockito.kotlin.mock
import ru.kiseru.checkers.domain.room.Room
import ru.kiseru.checkers.domain.room.RoomRepository
import ru.kiseru.checkers.room.impl.RoomServiceImpl

@ExtendWith(MockitoExtension::class)
class RoomServiceTest {

    @InjectMocks
    lateinit var underTest: RoomServiceImpl

    @Mock
    lateinit var roomRepository: RoomRepository

    @Test
    fun `findOrCreateRoom test when room doesn't exist`() {
        // when
        val actual = underTest.findOrCreateRoomById(1)

        // then
        assertThat(actual).isNotNull()
        assertThat(actual.id).isEqualTo(1)
    }

    @Test
    fun `findOrCreateRoom test when room exists`() {
        // given
        val room = mock<Room>()
        
        given { roomRepository.findRoom(eq(1)) }
            .willReturn(room)

        // when
        val actual = underTest.findOrCreateRoomById(1)

        // then
        assertThat(actual).isSameAs(room)
    }
}
